src/cryptonote_core/service_node_list.cpp

```diff
@@ -321,9 +329,214 @@ namespace service_nodes
+            /// assign it to the starving swarm
+            swarm_to_snodes.at(swarm_id).push_back(sn_pk);
+        }
+
+        if (!can_continue) break;
+      }
+
+    }
+
+    /// 3. Proceed to any "unsaturated" swarms (with fewer than max nodes)
+    {
+      while (!swarm_buffer.empty() && !all_swarms.empty()) {
+        // find the smallest -> fill it in first
+        const swarm_id_t smallest_swarm = *std::min_element(all_swarms.begin(), all_swarms.end(), cmp_swarm_sizes);
+
+        std::vector<crypto::public_key>& swarm = swarm_to_snodes.at(smallest_swarm);
+
+        if (swarm.size() == MAX_SWARM_SIZE) break;
+
+        const auto sn_pk = pop_random_snode(mersenne_twister, swarm_buffer);
+        swarm.push_back(sn_pk);
+      }
+    }
+
+    /// 4. If there are still enough nodes for MAX_SWARM_SIZE + some safety buffer, create a new swarm
+    while (swarm_buffer.size() >= MAX_SWARM_SIZE + SWARM_BUFFER) {
+
+      /// shuffle the queue and select MAX_SWARM_SIZE last elements
+      const auto new_swarm_id = get_new_swarm_id(mersenne_twister, all_swarms);
+
+      loki_shuffle(swarm_buffer, seed + new_swarm_id);
+
+      std::vector<crypto::public_key> selected_snodes;
+
+      for (auto i = 0u; i < MAX_SWARM_SIZE; ++i) {
+
+        /// get next node from the buffer
+        const crypto::public_key fresh_snode = swarm_buffer.back();
+        swarm_buffer.pop_back();
+
+        /// Try replacing nodes in existing swarms
+        if (swarm_to_snodes.size() > 0) {
+          /// a. Select a random swarm
+          const uint64_t swarm_idx = uniform_distribution_portable(mersenne_twister, swarm_to_snodes.size());
+          auto it = swarm_to_snodes.begin();
+          std::advance(it, swarm_idx);
+          std::vector<crypto::public_key>& selected_swarm = it->second;
+
+          /// b. Select a random snode
+          const crypto::public_key snode = pop_random_snode(mersenne_twister, selected_swarm);
+
+          /// c. Swap that node with a node in the queue, the old node will form a new swarm
+          selected_snodes.push_back(snode);
+          selected_swarm.push_back(fresh_snode);
+        } else {
+          /// If there are no existing swarms, create the first swarm directly from the queue
+          selected_snodes.push_back(fresh_snode);
+        }
+      }
+
+      swarm_to_snodes.insert({new_swarm_id, std::move(selected_snodes)});
+    }
+
+    /// 5. If there is a swarm with less than MIN_SWARM_SIZE, decommission that swarm (should almost never happen due to the safety buffer).
+    for (auto entry : swarm_to_snodes) {
+      if (entry.second.size() < MIN_SWARM_SIZE) {
+        LOG_PRINT_L1("swarm " << entry.first << " is DECOMMISSIONED");
+        /// TODO: move data to other swarms, then put snodes back in the queue
+      }
+    }
+
+    /// 6. Put nodes from the buffer back to the "buffer agnostic" data structure
+    swarm_to_snodes.insert({QUEUE_SWARM_ID, std::move(swarm_buffer)});
```

Comment:
if you take a reference to swarm_buffer, do we need this last move step?

Reply:
The map becomes the owner of the vector, so the move/copy is needed.